### ISSUE TITLE
Add methods to support computing on GPU

### ISSUE THREAD

**Issue by AlanAboudib** [OpenMined/PySyft#3615]
## Feature Description

In PyTorch, we can use `Tensor.to(<device>)` and `DataParallel` to choose between performing computations on either the CPU or on GPU(s).

I suggest to add similar functionality to PySyft to allow for performing computation on GPU. However, this has some subtleties:

1- When working in SMPC mode, each tensor is broken into multiple tensors with the same size. We call them shares. Each share lives on a different worker. So on each such worker we will have a share for a batch of training data, and a share of the tensor representing network parameters:

```
a_enc = a.share(bob, alice, crypto_provider = crypto_provider)
```

We should be able to call something like

```
a_enc = a_enc.to("cuda") 
```

which in turns, will load the corresponding share on the workers GPU if one is available. If no GPU is available on a given worker, we can revert to CPU or raise an error message. This could be controlled by an argument like:

```
a_enc = a_enc.to("cuda", allow_revert = True/False)
```

2- In FL mode, we could for example control this directly in the `send()` method:

```
a = a.send("bob", device = "cuda", allow_revert = True/False)
```

or

```
a = a.send("bob")
a = a.to("cuda", allow_revert = True/False)
```

This should allow for GPU acceleration especially when working with bigger models. Of course, network latency will be still an issue.

**Comment by karlhigley** [comment]
@AlanAboudib Those are good suggestions how the API might look. 👍 

Could you share an example of what it might look like on the model end? Want to make sure I'm following where the parallelism that GPUs can accelerate comes from.

**Comment by iamtrask** [comment]
@cereallarceny whats the new policy on new GitHub tags?

**Comment by cereallarceny** [comment]
Make whatever issue labels you want, but at the minimum you should assign a
type, status, priority, and severity issue label.

**Comment by AlanAboudib** [comment]
So acceleration comes from accelerating the forward pass and the backward pass to in 'loss.backward()'.

So for the model, it could look something like

```
# SMPC
model.share(bob, alice, crypto_provider = crypto_provider)
model = sy.DataParallel(model, max_devices = 'all'/<max allowed GPUs>)
model = model.to("cuda", allow_revert = True/False) 

# FL
model.send(bob)
model = sy.DataParallel(model, max_devices = 'all'/<max allowed GPUs>)
model = model.to("cuda", allow_revert = True/False) 
```

It should be something like that. But of course more subtlities will pop up.

Is that clear @karlhigley ?

**Comment by karlhigley** [comment]
I think I get the general idea, but perhaps it's the subtleties I've been missing. We have so many examples that demonstrate secret sharing with very small tensors that it was a bit of a mental leap to think about encrypted embedding layers (for example.)
